### PR TITLE
issue/3148 Added partly correct and any correct answer options to ItemsQuestionModel

### PR DESCRIPTION
**Pull request by oliverfoster**
#3148

Added support for selecting any correct answer.
Added support for marking as partly correct if any `_isPartlyCorrect` item is selected.

### Added
* `_isPartlyCorrect` to question items

### Changed
* `isCorrect` is `true` when a `_selectable` number of `_shouldBeSelected: true` items are active
* `isCorrect` is `false` if any `_isPartlyCorrect: true` items are active
* `isPartlyCorrect` is `true` if any `_isPartlyCorrect: true` items are active

To test in the AAT use https://github.com/adaptlearning/adapt-contrib-mcq/pull/160 and https://github.com/adaptlearning/adapt-contrib-gmcq/pull/130

### CODE REVIEW

File: src/core/js/models/itemsQuestionModel.js [hunk 1]
```diff
@@ -97,10 +109,12 @@ export default class ItemsQuestionModel extends BlendedItemsComponentQuestionMod
 
     this.set(props);
 
+    const hasAtLeastSelectableCorrectAnswers = (props._numberOfCorrectAnswers >= selectable);
     const hasRightNumberOfCorrectAnswers = (props._numberOfCorrectAnswers === props._numberOfRequiredAnswers);
     const hasNoIncorrectAnswers = !props._numberOfIncorrectAnswers;
+    const hasNoPartlyCorrectAnswers = !props._numberOfPartlyCorrectAnswers;
 
-    return hasRightNumberOfCorrectAnswers && hasNoIncorrectAnswers;
+    return (hasRightNumberOfCorrectAnswers || hasAtLeastSelectableCorrectAnswers) && hasNoIncorrectAnswers && hasNoPartlyCorrectAnswers;
```
Comment (eleanor-heath) [inline src/core/js/models/itemsQuestionModel.js:117]: suggestion: can we switch these `hasRightNumberOfCorrectAnswers || hasAtLeastSelectableCorrectAnswers` so they are in the same order as the order the variables are assigned. 
`return (hasAtLeastSelectableCorrectAnswers || hasRightNumberOfCorrectAnswers) && hasNoIncorrectAnswers && hasNoPartlyCorrectAnswers;`

File: src/core/js/models/itemsQuestionModel.js [hunk 2]
```diff
@@ -70,21 +70,33 @@ export default class ItemsQuestionModel extends BlendedItemsComponentQuestionMod
...
       if (!itemModel.get('_isActive')) return;
 
+      if (isPartlyCorrect) {
+        // force partly correct feedback and mark as correct
+        props._isAtLeastOneCorrectSelection = true;
+        props._numberOfPartlyCorrectAnswers++;
+        itemModel.set('_isCorrect', true);
+      }
+
       if (!itemShouldBeActive) {
```
Comment (eleanor-heath) [inline src/core/js/models/itemsQuestionModel.js:100]: could do 
 if (itemShouldBeActive) {
   props._numberOfRequiredAnswers++;
 } else {
  props._numberOfIncorrectAnswers++;
  return;
}

**Review by eleanor-heath — APPROVED** [review]

File: src/core/js/models/itemsQuestionModel.js (hunk 2 above)
Comment (oliverfoster) [reply src/core/js/models/itemsQuestionModel.js:100]: That'll not work quite as intended as the counts go out-of-wack. I'll rewrite to make it clearer.

**Review by eleanor-heath — APPROVED** [review]

**Review by cahirodoherty-learningpool — APPROVED** [review]
👀

File: src/core/js/models/itemsQuestionModel.js [hunk 3]
```diff
@@ -69,38 +69,36 @@ export default class ItemsQuestionModel extends BlendedItemsComponentQuestionMod
...
-        props._numberOfIncorrectAnswers++;
-        return;
-      }
-
-      props._isAtLeastOneCorrectSelection = true;
-      props._numberOfCorrectAnswers++;
-      itemModel.set('_isCorrect', true);
-    });
+    const numberOfSelectableAnswers = this.get('_selectable');
+    const hasAtLeastSelectableCorrectAnswers = (props._numberOfCorrectAnswers >= numberOfSelectableAnswers);
```
Comment (tomgreenfield) [inline src/core/js/models/itemsQuestionModel.js:93]: Please rename to be clearer as discussed.

**Review by guywillis — APPROVED** [review]
👀

**Pull request merged**